<?xml version="1.0" ?>
<sdf version="1.8">

    <world name="env_small">
        
        <plugin
            filename="ignition-gazebo-physics-system"
            name="ignition::gazebo::systems::Physics">
        </plugin>
        <plugin
            filename="ignition-gazebo-user-commands-system"
            name="ignition::gazebo::systems::UserCommands">
        </plugin>
        <plugin
            filename="ignition-gazebo-scene-broadcaster-system"
            name="ignition::gazebo::systems::SceneBroadcaster">
        </plugin>
        <plugin
            filename="ignition-gazebo-sensors-system"
            name="ignition::gazebo::systems::Sensors">
            <render_engine>ogre2</render_engine>
        </plugin>
        
        
        <physics name="1ms" type="ignored">
          <max_step_size>0.001</max_step_size>
          <real_time_factor>1.0</real_time_factor>
          <gravity>0 0 -9.8</gravity>
        </physics>

        


        <light type="directional" name="sun">
            <cast_shadows>true</cast_shadows>
            <pose>0 0 10 0 0 0</pose>
            <diffuse>0.8 0.8 0.8 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
            <attenuation>
                <range>1000</range>
                <constant>0.9</constant>
                <linear>0.01</linear>
                <quadratic>0.001</quadratic>
            </attenuation>
            <direction>-0.5 0.1 -0.9</direction>
        </light>

      <gui fullscreen="0">
      <!-- 3D scene -->
      <plugin filename="MinimalScene" name="3D View">
        <ignition-gui>
          <title>3D View</title>
          <property type="bool" key="showTitleBar">false</property>
          <property type="string" key="state">docked</property>
        </ignition-gui>

        <engine>ogre2</engine>
        <scene>scene</scene>
        <ambient_light>0.4 0.4 0.4</ambient_light>
        <background_color>0.8 0.8 0.8</background_color>
        <camera_pose>-6 0 6 0 0.5 0</camera_pose>
      </plugin>

      <!-- Plugins that add functionality to the scene -->
      <plugin filename="EntityContextMenuPlugin" name="Entity context menu">
        <ignition-gui>
          <property key="state" type="string">floating</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <plugin filename="GzSceneManager" name="Scene Manager">
        <ignition-gui>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <plugin filename="InteractiveViewControl" name="Interactive view control">
        <ignition-gui>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <plugin filename="CameraTracking" name="Camera Tracking">
        <ignition-gui>
          <property key="resizable" type="bool">false</property>
          <property key="width" type="double">5</property>
          <property key="height" type="double">5</property>
          <property key="state" type="string">floating</property>
          <property key="showTitleBar" type="bool">false</property>
        </ignition-gui>
      </plugin>
      <!-- World control -->
      <plugin filename="WorldControl" name="World control">
        <ignition-gui>
          <title>World control</title>
          <property type="bool" key="showTitleBar">false</property>
          <property type="bool" key="resizable">false</property>
          <property type="double" key="height">72</property>
          <property type="double" key="width">121</property>
          <property type="double" key="z">1</property>

          <property type="string" key="state">floating</property>
          <anchors target="3D View">
            <line own="left" target="left"/>
            <line own="bottom" target="bottom"/>
          </anchors>
        </ignition-gui>

        <play_pause>true</play_pause>
        <step>true</step>
        <start_paused>true</start_paused>
        <use_event>true</use_event>

      </plugin>

      <!-- World statistics -->
      <plugin filename="WorldStats" name="World stats">
        <ignition-gui>
          <title>World stats</title>
          <property type="bool" key="showTitleBar">false</property>
          <property type="bool" key="resizable">false</property>
          <property type="double" key="height">110</property>
          <property type="double" key="width">290</property>
          <property type="double" key="z">1</property>

          <property type="string" key="state">floating</property>
          <anchors target="3D View">
            <line own="right" target="right"/>
            <line own="bottom" target="bottom"/>
          </anchors>
        </ignition-gui>

        <sim_time>true</sim_time>
        <real_time>true</real_time>
        <real_time_factor>true</real_time_factor>
        <iterations>true</iterations>
      </plugin>

      <plugin filename="ImageDisplay" name="Image Display">
        <ignition-gui>
          <property key="state" type="string">docked</property>
        </ignition-gui>
        <topic>rgbd_camera/image</topic>
      </plugin>

      <plugin filename="ImageDisplay" name="Image Display">
        <ignition-gui>
          <property key="state" type="string">docked</property>
        </ignition-gui>
        <topic>rgbd_camera/depth_image</topic>
      </plugin>

      <plugin filename="ImageDisplay" name="Image Display">
        <ignition-gui>
          <property key="state" type="string">docked</property>
        </ignition-gui>
        <topic>semantic/colored_map</topic>
      </plugin>

      <plugin filename="ImageDisplay" name="Image Display">
        <ignition-gui>
          <property key="state" type="string">docked</property>
        </ignition-gui>
        <topic>semantic/labels_map</topic>
      </plugin>


      <!-- Inspector -->
      <plugin filename="ComponentInspector" name="Component inspector">
        <ignition-gui>
          <property type="string" key="state">docked</property>
        </ignition-gui>
      </plugin>

      <!-- Entity tree -->
      <plugin filename="EntityTree" name="Entity tree">
        <ignition-gui>
          <property type="string" key="state">docked</property>
        </ignition-gui>
      </plugin> 
    </gui>

    <model name="ground_plane">
      <static>true</static>
      <link name="link">
        <collision name="collision">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>200 200</size>
            </plane>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>100</mu>
                <mu2>50</mu2>
              </ode>
            </friction>
          </surface>
        </collision>
        <visual name="visual">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>100 100</size>
            </plane>
          </geometry>
          <material>
            <ambient>0.8 0.8 0.8 1</ambient>
            <diffuse>0.8 0.8 0.8 1</diffuse>
            <specular>0.8 0.8 0.8 1</specular>
          </material>
        </visual>
      </link>
    </model>  

    <include>
        <pose>3 7 0 0 0 3.14</pose>

        <uri>
        https://fuel.gazebosim.org/1.0/OpenRobotics/models/Office Chair
        <scale> 2 2 2 </scale>
        </uri>
        
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>1</label>
        </plugin>

    </include>
    
    <include>

        <pose>6 -3 3 0 0 -1.0</pose>
        <uri>
        https://fuel.gazebosim.org/1.0/makerspet/models/tv_65in_emissive
        </uri>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>2</label>
        </plugin>

    </include>


    <include>
        <pose>7 2 0 0 0 0</pose>
        <uri>
        https://fuel.gazebosim.org/1.0/OpenRobotics/models/Reflective table
        </uri>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>3</label>
        </plugin>
    </include>



    
    <!-- North Boundary Wall -->
    <model name="north_wall">
      <static>true</static>
      <pose>0 10 0.5 0 0 0</pose> <!-- Adjust position as needed -->
      <link name="link">
          <visual name="visual">
              <geometry>
                  <box>
                      <size>20 1 10</size> <!-- Adjust size as needed -->
                  </box>
              </geometry>
              <material>
                  <ambient>1 0 0 1</ambient> <!-- Red color -->
                  <diffuse>1 0 0 1</diffuse> <!-- Red color -->
                  <specular>0.5 0.5 0.5 1</specular> <!-- Specular component to add some shine -->
              </material>
          </visual>
          <collision name="collision">
              <geometry>
                  <box>
                      <size>20 1 10</size> <!-- Adjust size as needed -->
                  </box>
              </geometry>
          </collision>
      </link>
      <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
          <label>4</label>
      </plugin>
    </model>

    <!-- South Boundary Wall -->
    <model name="south_wall">
        <static>true</static>
        <pose>0 -10 0.5 0 0 0</pose> <!-- Adjust position as needed -->
        <link name="link">
            <visual name="visual">
                <geometry>
                    <box>
                        <size>20 1 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
                <material>
                    <ambient>1 0 0 1</ambient> <!-- Red color -->
                    <diffuse>1 0 0 1</diffuse> <!-- Red color -->
                    <specular>0.5 0.5 0.5 1</specular> <!-- Specular component to add some shine -->
                </material>
            </visual>
            <collision name="collision">
                <geometry>
                    <box>
                        <size>20 1 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
            </collision>
        </link>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>4</label>
        </plugin>
    </model>

    <!-- East Boundary Wall -->
    <model name="east_wall">
        <static>true</static>
        <pose>10 0 0.5 0 0 0</pose> <!-- Adjust position as needed -->
        <link name="link">
            <visual name="visual">
                <geometry>
                    <box>
                        <size>1 20 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
                <material>
                    <ambient>1 0 0 1</ambient> <!-- Red color -->
                    <diffuse>1 0 0 1</diffuse> <!-- Red color -->
                    <specular>0.5 0.5 0.5 1</specular> <!-- Specular component to add some shine -->
                </material>
            </visual>
            <collision name="collision">
                <geometry>
                    <box>
                        <size>1 20 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
            </collision>
        </link>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>4</label>
        </plugin>
    </model>

    <!-- West Boundary Wall -->
    <model name="west_wall">
        <static>true</static>
        <pose>-10 0 0.5 0 0 0</pose> <!-- Adjust position as needed -->
        <link name="link">
            <visual name="visual">
                <geometry>
                    <box>
                        <size>1 20 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
                <material>
                    <ambient>1 0 0 1</ambient> <!-- Red color -->
                    <diffuse>1 0 0 1</diffuse> <!-- Red color -->
                    <specular>0.5 0.5 0.5 1</specular> <!-- Specular component to add some shine -->
                </material>
            </visual>
            <collision name="collision">
                <geometry>
                    <box>
                        <size>1 20 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
            </collision>
        </link>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>4</label>
        </plugin>
    </model>


    <model name="maze_1">
        <static>true</static>
        <pose>5 5 0.5 0 0 0</pose> <!-- Adjust position as needed -->
        <link name="link">
            <visual name="visual">
                <geometry>
                    <box>
                        <size>10 1 2</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
                <material>
                    <ambient>1 1 0 1</ambient> <!-- Red color -->
                    <diffuse>1 1 0 1</diffuse> <!-- Red color -->
                    <specular>0.5 0.5 0.5 1</specular> <!-- Specular component to add some shine -->
                </material>
            </visual>
            <collision name="collision">
                <geometry>
                    <box>
                        <size>10 1 2</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
            </collision>
        </link>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>5</label>
        </plugin>
    </model>

    <model name="maze_2">
        <static>true</static>
        <pose>5 5 0.5 0 0 0</pose> <!-- Adjust position as needed -->
        <link name="link">
            <visual name="visual">
                <geometry>
                    <box>
                        <size>1 10 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
                <material>
                    <ambient>1 1 0 1</ambient> <!-- Red color -->
                    <diffuse>1 1 0 1</diffuse> <!-- Red color -->
                    <specular>0.5 0.5 0.5 1</specular> <!-- Specular component to add some shine -->
                </material>
            </visual>
            <collision name="collision">
                <geometry>
                    <box>
                        <size>1 10 10</size> <!-- Adjust size as needed -->
                    </box>
                </geometry>
            </collision>
        </link>
        <plugin filename="ignition-gazebo-label-system" name="ignition::gazebo::systems::Label">
            <label>5</label>
        </plugin>
    </model>

    <include>
      <uri>husky.sdf</uri> <!-- Generalize the path-->
      <pose>-4 4 2 0 0 0</pose>
    </include>

  </world>


</sdf>













